src/blog/2023/07/how-to-build-a-secure-opc-ua-server-for-plcs-in-node-red.md

```diff
@@ -0,0 +1,642 @@
+You end up with a new flow that looks like the one below.
+
+![example-flow.png](./images/opc-ua-2/example-flow.png)
+All we care about here is the `Compact-Server` node.  In fact, we’ll just copy that node and paste it into the current flow we’ve been building.  Once we’ve copied the server node into our custom flow, we can discard the example flow.  The whole purpose of this was to simply populate the `address space` of the `Compact-Server` node with template code that will trivialize the programming for our custom application.
+
+Our custom flow should now look something like this.
+
+![flow-with-compact-server.png](./images/opc-ua-2/flow-with-compact-server.png)
+- note - I’ve added some comments to make the flow even easier to follow.  Similar to commenting code, commenting flows is good practice.
+
+Open up the `Compact-Server` node and jump straight to the address space.  
+
+![compact-server-node-address-space.png](./images/opc-ua-2/compact-server-node-address-space.png)
+- note - we won’t go into detail on what the address space actually is in this article, or the details of the `Compact-Server` node, as it was covered in [part 1 of this OPC UA series](https://flowforge.com/blog/2023/07/how-to-deploy-a-basic-opc-ua-server-in-node-red/).  Please read that article if you are unfamiliar with it.
+
+There are 4 key things we’ll modify in the address space template code - 
+
+1. Bring in our `conveyorData` context variables 
+2. create our custom folder structure
+3. define our context variables as OPC UA nodes
+4. create custom browser views for our nodes
+
+### Bring in Context Variables <a name="context-variables"></a>
+
+Starting from the section of code where it’s bringing in the context variables defined in the example, delete that code
+
+```javascript 
+  this.sandboxFlowContext.set("isoInput1", 0);
+  this.setInterval(() => {
+    flexServerInternals.sandboxFlowContext.set(
+      "isoInput1",
+      Math.random() + 50.0
+    );
+  }, 500);
+  this.sandboxFlowContext.set("isoInput2", 0);
+ ...
+ this.sandboxFlowContext.set("isoOutput8", 0);
+```
+
+and replace it with our `conveyorData` context variables.
+
+```javascript 
+  this.sandboxFlowContext.set("conveyorData.Conveyor_RTS", false);
+  this.sandboxFlowContext.set("conveyorData.Robot_RTS", false);
+  this.sandboxFlowContext.set("conveyorData.Robot_Position", 0);
+  this.sandboxFlowContext.set("conveyorData.Conveyor_Running", false);
+  this.sandboxFlowContext.set("conveyorData.Line4_State", 0);
+  this.sandboxFlowContext.set("conveyorData.Line4_Fault", false);
+```
+
+### Create Custom Folder Structure <a name="custom-folder-structure"></a>
+
+Starting from the section of code where the example folder structure is defined, delete it and replace it with our custom folder structure defined in our architecture - 
+
+![opc-folder-structure.png](./images/opc-ua-2/opc-folder-structure.png)
+Delete the section of code starting from here - 
```

Comment:
I find it really odd that you are deleting and adding code. Where did the added code come from? Is there an easier way to do this than delete the old code?

Reply:
If you go earlier in the article, we go through the effort to copy in the example server flow covered in the first opc article.  we're taking that "template" program and modifying it by modifying only relevant sections and replacing it with our own code.  if we were building the code from scratch this project would be a lot more complicated to figure out.